privilege/privileges/cache.go

```diff
@@ -0,0 +1,273 @@
+			if !ok {
+				return errInvalidPrivilegeType.Gen("Unknown Privilege Type!")
+			}
+			value.Privileges |= priv
+		}
+	}
+	p.User = append(p.User, value)
+	return nil
+}
+
+func (p *MySQLPrivilege) decodeDBTableRow(row *ast.Row, fs []*ast.ResultField) error {
+	var value dbTableRow
+	for i, f := range fs {
+		d := row.Data[i]
+		switch {
+		case f.ColumnAsName.L == "user":
+			value.User = d.GetString()
+		case f.ColumnAsName.L == "host":
+			value.Host = d.GetString()
+		case f.ColumnAsName.L == "db":
+			value.DB = d.GetString()
+		case d.Kind() == types.KindMysqlEnum:
+			ed := d.GetMysqlEnum()
+			if ed.String() != "Y" {
+				continue
+			}
+			priv, ok := mysql.Col2PrivType[f.ColumnAsName.O]
+			if !ok {
+				return errInvalidPrivilegeType.Gen("Unknown Privilege Type!")
+			}
+			value.Privileges |= priv
+		}
+	}
+	p.DB = append(p.DB, value)
+	return nil
+}
+
+func (p *MySQLPrivilege) decodeTablesPrivTableRow(row *ast.Row, fs []*ast.ResultField) error {
+	var value tablesPrivTableRow
+	for i, f := range fs {
+		d := row.Data[i]
+		switch {
+		case f.ColumnAsName.L == "user":
+			value.User = d.GetString()
+		case f.ColumnAsName.L == "host":
+			value.Host = d.GetString()
+		case f.ColumnAsName.L == "db":
+			value.DB = d.GetString()
+		case f.ColumnAsName.L == "table_name":
+			value.TableName = d.GetString()
+		case f.ColumnAsName.L == "table_priv":
+			priv, err := decodeSetToPrivilege(d.GetMysqlSet())
+			if err != nil {
+				return errors.Trace(err)
+			}
+			value.TablePriv = priv
+		case f.ColumnAsName.L == "column_priv":
+			priv, err := decodeSetToPrivilege(d.GetMysqlSet())
+			if err != nil {
+				return errors.Trace(err)
+			}
+			value.ColumnPriv = priv
+		}
+	}
+	p.TablesPriv = append(p.TablesPriv, value)
+	return nil
+}
+
+func (p *MySQLPrivilege) decodeColumnsPrivTableRow(row *ast.Row, fs []*ast.ResultField) error {
+	var value columnsPrivTableRow
+	for i, f := range fs {
+		d := row.Data[i]
+		switch {
+		case f.ColumnAsName.L == "user":
+			value.User = d.GetString()
+		case f.ColumnAsName.L == "host":
+			value.Host = d.GetString()
+		case f.ColumnAsName.L == "db":
+			value.DB = d.GetString()
+		case f.ColumnAsName.L == "table_name":
+			value.TableName = d.GetString()
+		case f.ColumnAsName.L == "column_name":
+			value.ColumnName = d.GetString()
+		case f.ColumnAsName.L == "timestamp":
+			value.Timestamp, _ = d.GetMysqlTime().Time.GoTime(time.Local)
+		case f.ColumnAsName.L == "column_priv":
+			priv, err := decodeSetToPrivilege(d.GetMysqlSet())
+			if err != nil {
+				return errors.Trace(err)
+			}
+			value.ColumnPriv = priv
+		}
+	}
+	p.ColumnsPriv = append(p.ColumnsPriv, value)
+	return nil
+}
+
+func decodeSetToPrivilege(s types.Set) (mysql.PrivilegeType, error) {
+	var ret mysql.PrivilegeType
+	for _, str := range strings.Split(s.Name, ",") {
+		priv, ok := mysql.SetStr2Priv[str]
+		if !ok {
+			return ret, errInvalidPrivilegeType.Gen("Unknown PrivilegeType!")
```

Comment:
It should not happen, table schema of `user` `db` `tables_priv` is controlled by us.
If the program goes to this line, something must be wrong...
In one word, this not a general purpose error information.